core/src/main/java/org/elasticsearch/common/io/Streams.java

```diff
@@ -236,4 +241,57 @@ public static void readAllLines(InputStream input, Callback<String> callback) th
             }
         }
     }
+
+    /**
+     * Wraps the given {@link BytesStreamOutput} in a {@link StreamOutput} that simply flushes when
+     * close is called. The returned values from this method also implements the {@link BytesStream}
+     * interface.
+     */
+    public static StreamOutput flushOnCloseStream(BytesStreamOutput os) {
+        return new FlushOnCloseOutputStream(os);
+    }
+
+    /**
+     * A wrapper around a {@link BytesStreamOutput} that makes the close operation a flush. This is
+     * needed as sometimes a stream will be closed but the bytes that the stream holds still need
+     * to be used and the stream cannot be closed until the bytes have been consumed.
+     */
+    private static class FlushOnCloseOutputStream extends StreamOutput implements BytesStream {
```

Comment:
We implement BytesStream because on the rest layer we use these streams with an XContentBuilder and XContentBuilder expects that the stream is a BytesStream; it does a cast to BytesStream in the bytes method. This isn't clean, but we'd need another interface for the return type to be correct in what is being returned